test/load-generator/acme/directory.go

```diff
@@ -0,0 +1,208 @@
+// to fetch the raw bytes of the server's directory resource. If the URL is
+// invalid, if there is an error getting the directory bytes, or if the HTTP
+// response code is not 200 an error is returned.
+func getRawDirectory(directoryURL string) ([]byte, error) {
+	if directoryURL == "" {
+		return nil, ErrEmptyDirectory
+	}
+
+	if _, err := url.Parse(directoryURL); err != nil {
+		return nil, ErrInvalidDirectoryURL
+	}
+
+	httpClient := &http.Client{
+		Transport: &http.Transport{
+			DialContext: (&net.Dialer{
+				Timeout:   10 * time.Second,
+				KeepAlive: 30 * time.Second,
+			}).DialContext,
+			TLSHandshakeTimeout: 5 * time.Second,
+			TLSClientConfig: &tls.Config{
+				// Bypassing CDN or testing against Pebble instances can cause
+				// validation failures. For a **test-only** tool its acceptable to skip
+				// cert verification of the ACME server's HTTPs certificate.
+				InsecureSkipVerify: true,
+			},
+			MaxIdleConns:    1,
+			IdleConnTimeout: 15 * time.Second,
+		},
+		Timeout: 10 * time.Second,
+	}
+
+	resp, err := httpClient.Get(directoryURL)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, ErrInvalidDirectoryHTTPCode
+	}
+
+	rawDirectory, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return rawDirectory, nil
+}
+
+// NewDirectory creates a Directory populated from the ACME directory resource
+// returned by a GET request to the provided directoryURL. An error is returned
+// if the directory URL is invalid, or if there is a problem connecting to the
+// directory URL, unmarshaling the JSON directory resource, or if the directory
+// is missing required endpoints (see RequiredEndpoints).
+func NewDirectory(directoryURL string) (*Directory, error) {
+	// Fetch the raw directory JSON
+	dirContents, err := getRawDirectory(directoryURL)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal the directory
+	var dirResource map[string]interface{}
+	if err := json.Unmarshal(dirContents, &dirResource); err != nil {
+		return nil, ErrInvalidDirectoryJSON
+	}
+
+	// serverURL tries to find a valid url.URL for the provided endpoint in
+	// the unmarshaled directory resource.
+	serverURL := func(name Endpoint) (*url.URL, error) {
+		if rawURL, ok := dirResource[string(name)]; !ok {
+			return nil, ErrMissingEndpoint{endpoint: name}
+		} else if urlString, ok := rawURL.(string); !ok {
+			return nil, ErrInvalidEndpointURL{endpoint: name, value: urlString}
+		} else if url, err := url.Parse(urlString); err != nil {
+			return nil, ErrInvalidEndpointURL{endpoint: name, value: urlString}
+		} else {
+			return url, nil
+		}
+	}
+
+	// Create an empty directory to populate
+	directory := &Directory{
+		endpointURLs: make(map[Endpoint]*url.URL),
+	}
+
+	// Every required endpoint must have a valid URL populated from the directory
+	for _, endpointName := range RequiredEndpoints {
```

Comment:
I think we'll also need a way to extract the ToS from this also?

Reply:
Good point :+1: I'll add that.

Reply:
I realized this isn't actually needed now that the ACME v1 code is ripped out. The RFC 8555 based V2 code unconditionally sends `termsOfServiceAgreed: true` and doesn't need to know the ToS URL. 

I realized this _after_ I had already implemented it in 0f3e72d :sweat: I'm somewhat partial to leaving it in since I already wrote it but I can revert it if you'd prefer since there isn't a need for it in the `load-generator` codebase presently. LMK what you prefer.

Reply:
Hm, I think it's probably fine to leave. Other meta stuff might be useful to have down the line anyway.

Reply:
Sounds good :+1: